documentation/ImplementingASemanticCheck.md

```diff
@@ -0,0 +1,813 @@
+<!--===- documentation/FrontEndTutorial.md 
+  
+   Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+   See https://llvm.org/LICENSE.txt for license information.
+   SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+  
+-->
+# Introduction
+I recently added a semantic check to the f18 compiler front end.  This document
+describes my thought process and the resulting implementation.
+
+For more information about the compiler, start with the 
+[compiler overview](Overview.md).
+
+# Problem definition
+
+In the 2018 Fortran standard, section 11.1.7.4.3, paragraph 2, states that:
+
+```
+Except for the incrementation of the DO variable that occurs in step (3), the DO variable 
+shall neither be redefined nor become undefined while the DO construct is active.
+```
+One of the ways that DO variables might be redefined is if they are passed to
+functions with dummy arguments whose `INTENT` is `INTENT(OUT)` or
+`INTENT(INOUT)`.  I implemented this semantic check.  Specifically, I changed
+the compiler to emit an error message if an active DO variable was passed to a
+dummy argument of a FUNCTION with INTENT(OUT).  Similarly, I had the compiler
+emit a warning if an active DO variable was passed to a dummy argument with
+INTENT(INOUT).  Previously, I had implemented similar checks for SUBROUTINE
+calls.
+
+# Creating a test
+
+My first step was to create a test case to cause the problem.  I called it testfun.f90 and used it to check the behavior of other Fortran compilers.  Here's the initial version:
+
+```fortran
+  subroutine s()
+    Integer :: ivar, jvar
+
+    do ivar = 1, 10
+      jvar = intentOutFunc(ivar) ! Error since ivar is a DO variable
+    end do
+
+  contains
+    function intentOutFunc(dummyArg)
+      integer, intent(out) :: dummyArg
+      integer  :: intentOutFunc
+
+      dummyArg = 216
+    end function intentOutFunc
+  end subroutine s
+```
+
+I verified that other Fortran compilers produced an error message at the point
+of the call to `intentOutFunc()`:
+
+```fortran
+      jvar = intentOutFunc(ivar) ! Error since ivar is a DO variable
+```
+
+
+I also used this program to produce a parse tree for the program using the command:
+```bash
+  ./bin/f18 -fdebug-dump-parse-tree -fparse-only testfun.f90
```

Comment:
maybe just `f18` instead of `./bin/f18`?

Reply:
Yes.  I think your choice is better.